### PR TITLE
QR-Code: Add evaluation sample

### PR DESCRIPTION
**Pull request by APrigarina**
Sample for evaluation QR-code algorithm, includes markup for dataset

```
force_builders_only=Docs
```

### CODE REVIEW

**Comment by APrigarina** [comment]
@alalek , @allnes , @rayonnant14, please review

**Comment by asmorkalov** [comment]
@allnes @rayonnant14 Friendly reminder.

File: samples/python/evaluate_qrcode/markup/img_1080p_config.json [hunk 1]
```diff
@@ -0,0 +1,274 @@
+{
```
Comment (allnes) [inline samples/python/evaluate_qrcode/markup/img_1080p_config.json:1]: I think that configuration files can be moved to opencv_extra

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
+#!/usr/bin/env python
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:1]: I don't see OpenCV's copyright

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+
+        total_number_images = len(input_data['test_images'])
+
+        for idx, input_image in enumerate(input_images):
+            image_data = input_data['test_images'][idx]
+
+            path_to_image = os.path.join(path_to_category, input_image)
+            image_to_detect = cv2.imread(path_to_image)
+
+            if metric == 1:
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:71]: please, name metrics and remove digits

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+import os
+import sys
+from os.path import join
+
+import cv2
+import numpy as np
+import json
+
+
+def main():
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:23]: please change `def main():` to `if __name__ == "__main__":` cause its right form program's entry point.

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+and markup files located in markup/ folder
+for evaluation QR-code detection algorithm
+
+'''
+
+# Python 2/3 compatibility
+import optparse
+import os
+import sys
+from os.path import join
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:16]: please use pathlib instead os.path, all reasons mentioned [this](https://habr.com/ru/post/453862/)

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+import cv2
+import numpy as np
+import json
+
+
+def main():
+    parser = optparse.OptionParser()
+    parser.add_option('--images', default="qrcode-datasets/datasets", help="Location of directory with input images")
+    parser.add_option('--data', default="markup", help="Location of directory with input data")
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:26]: please add directory for label's files

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+    path_to_images = options.images
+    path_to_data = options.data
+
+    dir_images = [f for f in sorted(os.listdir(path_to_images))]
+    dir_data = [f for f in sorted(os.listdir(path_to_data))]
+
+    run_metric_through_categories(path_to_images, dir_images, path_to_data, dir_data, 1)
+    run_metric_through_categories(path_to_images, dir_images, path_to_data, dir_data, 2)
+    run_metric_through_categories(path_to_images, dir_images, path_to_data, dir_data, 3)
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:37]: please don't duplicate source code with same functionality

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+            scores.append(np.mean(category_score))
+
+    if metric <= 3:
+        print_result(categories, detection_stat, decoding_stat)
+    else:
+        print_result_scores(categories, scores)
+
+
+def get_input_images(path_to_images):
+    input_images = [f for f in sorted(os.listdir(path_to_images)) if f.endswith("jpg") or f.endswith("JPG") or f.endswith("png")]
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:111]: I'm think that can be redundant.

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+def get_input_data(path_to_data, input_file):
+    path_to_file = os.path.join(path_to_data, input_file)
+    json_file = open(path_to_file, 'r')
+    input_data = json.load(json_file)
+
+    return input_data
+
+
+def print_metric_description(metric):
+    if metric == 1:
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:125]: name metrics, too

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+from os.path import join
+
+import cv2
+import numpy as np
+import json
+
+
+def main():
+    parser = optparse.OptionParser()
+    parser.add_option('--images', default="qrcode-datasets/datasets", help="Location of directory with input images")
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:25]: please change option as require parameters

File: samples/python/evaluate_qrcode/evaluate_qrcode.py (hunk 8 above)
Comment (allnes) [reply samples/python/evaluate_qrcode/evaluate_qrcode.py:111]: I don't understand, how has it used?

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+http://www.fit.vutbr.cz/research/groups/graph/pclines/pub_page.php?id=2012-SCCG-QRtiles
+and markup files located in markup/ folder
+for evaluation QR-code detection algorithm
+
+'''
+
+# Python 2/3 compatibility
+import optparse
+import os
+import sys
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:15]: unused library

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+        detected_in_category = 0
+        decoded_in_category = 0
+
+        path_to_category = os.path.join(path_to_images, category)
+        number_qrcodes_in_category = get_number_qrcodes_in_category(path_to_data, dir_data[idx])
+
+        input_images = get_input_images(path_to_category)
+        input_data = get_input_data(path_to_data, dir_data[idx])
+
+        total_number_images = len(input_data['test_images'])
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:63]: what kind of type does `input_data` have? if `np.array` then please use `input_data.size`

File: samples/python/evaluate_qrcode/evaluate_qrcode.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+                decoded +=1
+
+    return detected, decoded
+
+
+def compute_metric_2(image_to_detect, qrDetector):
+    detected = 0
+    decoded = 0
+
+    ok, bboxes = qrDetector.detectMulti(image_to_detect)
```
Comment (allnes) [inline samples/python/evaluate_qrcode/evaluate_qrcode.py:192]: what does `ok` variable mean?

File: samples/python/evaluate_qrcode/evaluate_qrcode.py (hunk 3 above)
Comment (APrigarina) [reply samples/python/evaluate_qrcode/evaluate_qrcode.py]: thanks, renamed

File: samples/python/evaluate_qrcode/evaluate_qrcode.py (hunk 5 above)
Comment (APrigarina) [reply samples/python/evaluate_qrcode/evaluate_qrcode.py]: thanks! fixed it

File: samples/python/evaluate_qrcode/evaluate_qrcode.py (hunk 13 above)
Comment (tweenietomatoes) [reply samples/python/evaluate_qrcode/evaluate_qrcode.py]: quirc is updated as well, current qr code decoder is buggy especially using video capture.

Comment (allnes) [reply samples/python/evaluate_qrcode/evaluate_qrcode.py]: > quirc is updated as well, current qr code decoder is buggy especially using video capture.

thank you, we'll try update quirc

**Comment by asmorkalov** [comment]
@APrigarina friendly reminder.

**Comment by asmorkalov** [comment]
@APrigarina @allnes Friendly reminder.

**Comment by asmorkalov** [comment]
@APrigarina @allnes Friendly reminder.

**Comment by asmorkalov** [comment]
@APrigarina @allnes Friendly reminder.

**Comment by asmorkalov** [comment]
@APrigarina @allnes Do you have any progress with the patch?

**Comment by alalek** [comment]
This code will be moved to `opencv_benchmarks` repository.

**Pull request closed without merging**